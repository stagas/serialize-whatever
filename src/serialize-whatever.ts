import { Ctor } from 'everyday-types'

type AnyFn = (x: any) => any

type Entry = {
  $r: number // ref id
  $t: string // type
  $v: any // value
}

export type Ptr = {
  $p: number // pointer to reference id
}

const trailingNumbersRegExp = /\d+$/g
const stripTrailingNumbers = (x: string) => x.replace(trailingNumbersRegExp, '')

export const thru = (x: any) => x

export const createContext = (extraTypes: [Ctor, [AnyFn, AnyFn]][] = [], ptrs = new Map<object, Ptr>()) => {
  const Types = new Map([
    [Object],
    [Array, [thru, thru]],
    [Map, [x => [...x]]],
    [Set, [x => [...x]]],
    [Date],
    ...extraTypes,
  ] as [Ctor, [AnyFn, AnyFn]][])

  const TypesMap = [...Types.keys()].map(x => [stripTrailingNumbers(x.name), x])

  let pointer = 0

  const defaultSerializer = (obj: object & { toJSON?: () => object }): any =>
    obj.toJSON ? obj.toJSON() : Object.assign({}, obj)

  const createEntry = (serializer: AnyFn, obj: object & { toJSON?: () => object }): Entry | Ptr =>
    ptrs.get(obj) ?? ({
      $r: (ptrs.set(obj, { $p: ++pointer } as Ptr), pointer),
      $t: stripTrailingNumbers(obj.constructor.name),
      $v: serializer(obj),
    })

  const replacer = (
    top: unknown,
    clear = true,
  ) => (clear && (pointer = 0, ptrs.clear()), function (this: any, key: string): string | Entry | Ptr {
    const value = this[key]

    if (
      (value === top && key !== '')
      || (typeof value === 'object' && value.$r)
      || this.$r
    ) {
      return value
    }

    if (typeof value === 'object') {
      const [serializer = defaultSerializer] = Types.get(value.constructor) ?? []
      return createEntry(serializer, value)
    }

    return value
  })

  const reviver = (
    classes: Ctor[],
    refs: Map<number, object> = new Map(),
    pending: Map<number, Set<[any, string | ((result: any) => void)]>> = new Map(),
  ) => {
    const types = new Map([
      ...TypesMap,
      ...classes.map(x => [stripTrailingNumbers(x.name), x]),
    ] as [string, Ctor][])

    const getRef = (value: any, key: any, owner: any) => {
      const { $p } = value
      if (refs.has($p)) return refs.get($p)
      else {
        let queued = pending.get($p)
        if (!queued) pending.set($p, queued = new Set())
        queued.add([owner, key])
        return value
      }
    }

    return function (this: any, key: string, value: any) {
      if (typeof value === 'object' && value !== null) {
        if (value.$t) {
          const { $r, $t, $v } = value as Entry
          const ctor = types.get($t)
          if (!ctor) {
            throw new TypeError(
              'Unable to deserialize object type: ' + $t
              + `\n  Be sure to pass the class like so:\n\n    deserialize(serialized, [${$t}])`
            )
          }
          const [, deserializer = ((x: any) => new ctor(x))] = Types.get(ctor) ?? []

          const result = deserializer($v)

          if ($t === 'Map') {
            for (const pv of pending.values()) {
              for (const q of pv) {
                if ($v.includes(q[0])) {
                  if (q[1] == '0') {
                    q[1] = (key: any) => {
                      const p = q[0][0]
                      const v = result.get(p)
                      result.set(key, v)
                      result.delete(p)
                    }
                  } else if (q[1] == '1') {
                    q[1] = (value: any) => {
                      let p = q[0][0]
                      if (p?.$p) {
                        p = refs.get(p.$p)
                      }
                      result.set(p, value)
                    }
                  }
                }
              }
            }
          }

          if ($t === 'Set') {
            for (const pv of pending.values()) {
              for (const q of pv) {
                if ($v.includes(q[0][0])) {
                  const p = q[0][q[1] as any]
                  q[1] = (value: any) => {
                    result.add(value)
                    result.delete(p)
                  }
                }
              }
            }
          }

          refs.set($r, result)
          if (pending.has($r)) {
            pending.get($r)!.forEach(([parent, key]) => {
              if (typeof key === 'function') {
                key(result)
              } else {
                parent[key] = result
              }
            })
          }
          return result
        } else if (value.$p) {
          return getRef(value, key, this)
        }
      }
      return value
    }
  }

  return { replacer, reviver }
}

export const { replacer, reviver } = createContext()

export const serialize = (any: unknown, indent?: number, clear?: boolean) =>
  JSON.stringify(any, replacer(any, clear), indent)

export const deserialize = (
  serialized: string,
  classes: Ctor[] = [],
): unknown => JSON.parse(serialized, reviver(classes))
