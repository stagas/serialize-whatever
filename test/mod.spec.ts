import { deserialize, serialize } from '../src/serialize-whatever'
import { Cable, Plug } from './plugs-and-cables'

describe('serialize/deserialize', () => {
  it('works', () => {
    class Foo {
      a = 1
      b = 2
      constructor(state: Partial<Foo> = {}) {
        Object.assign(this, state)
      }
      toJSON() {
        return {
          a: this.a,
        }
      }
    }
    const obj = {
      a: 123,
      b: new Foo({ a: 2 }),
      c: new Date(123),
    }
    const str = serialize(obj)
    expect(str).toMatchSnapshot()
    const res = deserialize(str, [Foo])
    expect(res).toMatchSnapshot()
    expect(res).toEqual(obj)
  })

  it('preserves references', () => {
    class Foo {
      a = 1
      b = 2
      constructor(state: Partial<Foo> = {}) {
        Object.assign(this, state)
      }
      toJSON() {
        return {
          a: this.a,
        }
      }
    }
    const foo = new Foo({ a: 2 })
    const obj = {
      a: 123,
      b: foo,
      c: new Date(123),
      d: foo,
    }
    const str = serialize(obj)
    expect(str).toMatchSnapshot()
    const res = deserialize(str, [Foo]) as typeof obj
    expect(res).toMatchSnapshot()
    expect(res).toEqual(obj)
    expect(res.d).toBe(res.b)
  })

  it('common types', () => {
    const obj = {
      a: 123,
      b: new Date(123),
      c: new Map([[1, 2], [3, 4]]),
      d: new Set([1, 2, 3, 4]),
    }
    const str = serialize(obj)
    expect(str).toMatchSnapshot()
    const res = deserialize(str)
    expect(res).toMatchSnapshot()
    expect(res).toEqual(obj)
  })

  it('top level map', () => {
    const obj = new Map([[1, 2], [3, 4]])
    const str = serialize(obj)
    expect(str).toMatchSnapshot()
    const res = deserialize(str)
    expect(res).toMatchSnapshot()
    expect(res).toEqual(obj)
  })

  it('refs in maps and sets', () => {
    const foo = { hello: 'world' }
    const arr = [1, 2, 3]
    const arrRef = [1, 2, foo]
    const obj = {
      a: foo,
      b: new Date(123),
      c: new Map([[1, foo], [2, arr]] as any),
      d: new Set([1, 2, foo, 4]),
      e: arr,
      f: arrRef,
    }
    const str = serialize(obj)
    expect(str).toMatchSnapshot()
    const res = deserialize(str) as typeof obj
    expect(res).toMatchSnapshot()
    expect(res).toEqual(obj)
    expect(res.c.get(1)).toBe(res.a)
    expect(res.c.get(2)).toBe(res.e)
    expect(res.f[2]).toBe(res.a)
  })

  it('refs in maps keys', () => {
    const foo = { hello: 'world' }
    const bar = { one: 'two' }
    const obj = {
      a: new Map([[foo, bar]]),
    }
    const str = serialize(obj)
    expect(str).toMatchSnapshot()
    const res = deserialize(str) as typeof obj
    expect(res).toMatchSnapshot()
    expect(res).toEqual(obj)
  })

  it('cross refs', () => {
    const foo = { hello: 'world', bar: undefined } as any
    const bar = { one: 'two', b: foo }
    foo.bar = bar
    const obj = {
      a: foo,
      b: bar,
    } as any
    obj.c = obj.a
    obj.d = { a: obj.a, b: obj.b }
    const str = serialize(obj)
    expect(str).toMatchSnapshot()
    const res = deserialize(str)
    expect(res).toMatchSnapshot()
    expect(res).toEqual(obj)
  })

  it('array', () => {
    const obj = [{
      a: 123,
      b: new Date(123),
      c: new Map([[1, 2], [3, 4]]),
      d: new Set([1, 2, 3, 4]),
    }]
    const str = serialize(obj)
    expect(str).toMatchSnapshot()
    const res = deserialize(str)
    expect(res).toMatchSnapshot()
    expect(res).toEqual(obj)
  })

  it('plugs and cables', () => {
    const output = new Plug(Plug.Output, 'audio')
    const input = new Plug(Plug.Input, 'audio')
    output.connect(input)
    const obj: Record<'input' | 'output', Plug> = {
      output,
      input,
    }
    const str = serialize(obj)
    expect(str).toMatchSnapshot()
    const res = deserialize(str, [Plug, Cable])
    expect(res).toMatchSnapshot()
    expect(res).toEqual(obj)
  })
})
