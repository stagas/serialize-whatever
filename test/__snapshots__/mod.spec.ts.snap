
exports[`serialize/deserialize works 1`] = `"{\\"$r\\":1,\\"$t\\":\\"Object\\",\\"$v\\":{\\"a\\":123,\\"b\\":{\\"$r\\":2,\\"$t\\":\\"Foo\\",\\"$v\\":{\\"a\\":2}},\\"c\\":{\\"$r\\":3,\\"$t\\":\\"Date\\",\\"$v\\":\\"1970-01-01T00:00:00.123Z\\"}}}"`;

exports[`serialize/deserialize works 2`] = `
Object {
  "a": 123,
  "b": Object {
    "a": 2,
  },
  "c": 1970-01-01T00:00:00.123Z,
}
`;

exports[`serialize/deserialize preserves references 1`] = `"{\\"$r\\":1,\\"$t\\":\\"Object\\",\\"$v\\":{\\"a\\":123,\\"b\\":{\\"$r\\":2,\\"$t\\":\\"Foo\\",\\"$v\\":{\\"a\\":2}},\\"c\\":{\\"$r\\":3,\\"$t\\":\\"Date\\",\\"$v\\":\\"1970-01-01T00:00:00.123Z\\"},\\"d\\":{\\"$p\\":2}}}"`;

exports[`serialize/deserialize preserves references 2`] = `
Object {
  "a": 123,
  "b": Object {
    "a": 2,
  },
  "c": 1970-01-01T00:00:00.123Z,
  "d": Object {
    "a": 2,
  },
}
`;

exports[`serialize/deserialize common types 1`] = `"{\\"$r\\":1,\\"$t\\":\\"Object\\",\\"$v\\":{\\"a\\":123,\\"b\\":{\\"$r\\":2,\\"$t\\":\\"Date\\",\\"$v\\":\\"1970-01-01T00:00:00.123Z\\"},\\"c\\":{\\"$r\\":3,\\"$t\\":\\"Map\\",\\"$v\\":[{\\"$r\\":4,\\"$t\\":\\"Array\\",\\"$v\\":[1,2]},{\\"$r\\":5,\\"$t\\":\\"Array\\",\\"$v\\":[3,4]}]},\\"d\\":{\\"$r\\":6,\\"$t\\":\\"Set\\",\\"$v\\":[1,2,3,4]}}}"`;

exports[`serialize/deserialize common types 2`] = `
Object {
  "a": 123,
  "b": 1970-01-01T00:00:00.123Z,
  "c": Map {
    1 => 2,
    3 => 4,
  },
  "d": Set {
    1,
    2,
    3,
    4,
  },
}
`;

exports[`serialize/deserialize top level map 1`] = `"{\\"$r\\":1,\\"$t\\":\\"Map\\",\\"$v\\":[{\\"$r\\":2,\\"$t\\":\\"Array\\",\\"$v\\":[1,2]},{\\"$r\\":3,\\"$t\\":\\"Array\\",\\"$v\\":[3,4]}]}"`;

exports[`serialize/deserialize top level map 2`] = `
Map {
  1 => 2,
  3 => 4,
}
`;

exports[`serialize/deserialize refs in maps and sets 1`] = `"{\\"$r\\":1,\\"$t\\":\\"Object\\",\\"$v\\":{\\"a\\":{\\"$r\\":2,\\"$t\\":\\"Object\\",\\"$v\\":{\\"hello\\":\\"world\\"}},\\"b\\":{\\"$r\\":3,\\"$t\\":\\"Date\\",\\"$v\\":\\"1970-01-01T00:00:00.123Z\\"},\\"c\\":{\\"$r\\":4,\\"$t\\":\\"Map\\",\\"$v\\":[{\\"$r\\":5,\\"$t\\":\\"Array\\",\\"$v\\":[1,{\\"$p\\":2}]},{\\"$r\\":6,\\"$t\\":\\"Array\\",\\"$v\\":[2,{\\"$r\\":7,\\"$t\\":\\"Array\\",\\"$v\\":[1,2,3]}]}]},\\"d\\":{\\"$r\\":8,\\"$t\\":\\"Set\\",\\"$v\\":[1,2,{\\"$p\\":2},4]},\\"e\\":{\\"$p\\":7},\\"f\\":{\\"$r\\":9,\\"$t\\":\\"Array\\",\\"$v\\":[1,2,{\\"$p\\":2}]}}}"`;

exports[`serialize/deserialize refs in maps and sets 2`] = `
Object {
  "a": Object {
    "hello": "world",
  },
  "b": 1970-01-01T00:00:00.123Z,
  "c": Map {
    1 => [Circular],
    2 => Array [
      1,
      2,
      3,
    ],
  },
  "d": Set {
    1,
    2,
    [Circular],
    4,
  },
  "e": Array [
    1,
    2,
    3,
  ],
  "f": Array [
    1,
    2,
    [Circular],
  ],
}
`;

exports[`serialize/deserialize refs in maps keys 1`] = `"{\\"$r\\":1,\\"$t\\":\\"Object\\",\\"$v\\":{\\"a\\":{\\"$r\\":2,\\"$t\\":\\"Map\\",\\"$v\\":[{\\"$r\\":3,\\"$t\\":\\"Array\\",\\"$v\\":[{\\"$r\\":4,\\"$t\\":\\"Object\\",\\"$v\\":{\\"hello\\":\\"world\\"}},{\\"$r\\":5,\\"$t\\":\\"Object\\",\\"$v\\":{\\"one\\":\\"two\\"}}]}]}}}"`;

exports[`serialize/deserialize refs in maps keys 2`] = `
Object {
  "a": Map {
    Object {
      "hello": "world",
    } => Object {
      "one": "two",
    },
  },
}
`;

exports[`serialize/deserialize cross refs 1`] = `"{\\"$r\\":1,\\"$t\\":\\"Object\\",\\"$v\\":{\\"a\\":{\\"$r\\":2,\\"$t\\":\\"Object\\",\\"$v\\":{\\"hello\\":\\"world\\",\\"bar\\":{\\"$r\\":3,\\"$t\\":\\"Object\\",\\"$v\\":{\\"one\\":\\"two\\",\\"b\\":{\\"$p\\":2}}}}},\\"b\\":{\\"$p\\":3},\\"c\\":{\\"$p\\":2},\\"d\\":{\\"$r\\":4,\\"$t\\":\\"Object\\",\\"$v\\":{\\"a\\":{\\"$p\\":2},\\"b\\":{\\"$p\\":3}}}}}"`;

exports[`serialize/deserialize cross refs 2`] = `
Object {
  "a": Object {
    "bar": Object {
      "b": [Circular],
      "one": "two",
    },
    "hello": "world",
  },
  "b": Object {
    "b": [Circular],
    "one": "two",
  },
  "c": Object {
    "bar": [Circular],
    "hello": "world",
  },
  "d": Object {
    "a": [Circular],
    "b": [Circular],
  },
}
`;

exports[`serialize/deserialize array 1`] = `"{\\"$r\\":1,\\"$t\\":\\"Array\\",\\"$v\\":[{\\"$r\\":2,\\"$t\\":\\"Object\\",\\"$v\\":{\\"a\\":123,\\"b\\":{\\"$r\\":3,\\"$t\\":\\"Date\\",\\"$v\\":\\"1970-01-01T00:00:00.123Z\\"},\\"c\\":{\\"$r\\":4,\\"$t\\":\\"Map\\",\\"$v\\":[{\\"$r\\":5,\\"$t\\":\\"Array\\",\\"$v\\":[1,2]},{\\"$r\\":6,\\"$t\\":\\"Array\\",\\"$v\\":[3,4]}]},\\"d\\":{\\"$r\\":7,\\"$t\\":\\"Set\\",\\"$v\\":[1,2,3,4]}}}]}"`;

exports[`serialize/deserialize array 2`] = `
Array [
  Object {
    "a": 123,
    "b": 1970-01-01T00:00:00.123Z,
    "c": Map {
      1 => 2,
      3 => 4,
    },
    "d": Set {
      1,
      2,
      3,
      4,
    },
  },
]
`;

exports[`serialize/deserialize plugs and cables 1`] = `"{\\"$r\\":1,\\"$t\\":\\"Object\\",\\"$v\\":{\\"output\\":{\\"$r\\":2,\\"$t\\":\\"Plug\\",\\"$v\\":{\\"plugKind\\":\\"output\\",\\"cableKind\\":\\"audio\\",\\"cables\\":{\\"$r\\":3,\\"$t\\":\\"Map\\",\\"$v\\":[{\\"$r\\":4,\\"$t\\":\\"Array\\",\\"$v\\":[{\\"$r\\":5,\\"$t\\":\\"Cable\\",\\"$v\\":{\\"id\\":0,\\"plugs\\":{\\"$r\\":6,\\"$t\\":\\"Set\\",\\"$v\\":[{\\"$p\\":2},{\\"$r\\":7,\\"$t\\":\\"Plug\\",\\"$v\\":{\\"plugKind\\":\\"input\\",\\"cableKind\\":\\"audio\\",\\"cables\\":{\\"$r\\":8,\\"$t\\":\\"Map\\",\\"$v\\":[{\\"$r\\":9,\\"$t\\":\\"Array\\",\\"$v\\":[{\\"$p\\":5},{\\"$p\\":2}]}]}}}]},\\"outputCh\\":0,\\"inputCh\\":0,\\"gain\\":1}},{\\"$p\\":7}]}]}}},\\"input\\":{\\"$p\\":7}}}"`;

exports[`serialize/deserialize plugs and cables 2`] = `
Object {
  "input": Plug {
    "cableKind": "audio",
    "cables": Map {
      Cable {
        "gain": 1,
        "id": 0,
        "inputCh": 0,
        "outputCh": 0,
        "plugs": Set {
          [Circular],
          Plug {
            "cableKind": "audio",
            "cables": Map {
              [Circular] => [Circular],
            },
            "plugKind": "output",
          },
        },
      } => Plug {
        "cableKind": "audio",
        "cables": Map {
          [Circular] => [Circular],
        },
        "plugKind": "output",
      },
    },
    "plugKind": "input",
  },
  "output": Plug {
    "cableKind": "audio",
    "cables": Map {
      Cable {
        "gain": 1,
        "id": 0,
        "inputCh": 0,
        "outputCh": 0,
        "plugs": Set {
          [Circular],
          [Circular],
        },
      } => [Circular],
    },
    "plugKind": "output",
  },
}
`;

exports[`serialize/deserialize quirky 1`] = `"{\\"$r\\":1,\\"$t\\":\\"Object\\",\\"$v\\":{\\"a\\":{\\"$r\\":2,\\"$t\\":\\"Map\\",\\"$v\\":[{\\"$r\\":3,\\"$t\\":\\"Array\\",\\"$v\\":[{\\"$r\\":4,\\"$t\\":\\"Set\\",\\"$v\\":[{\\"$r\\":5,\\"$t\\":\\"Object\\",\\"$v\\":{\\"hello\\":\\"world\\",\\"map\\":{\\"$p\\":2}}},{\\"$p\\":2}]},{\\"$p\\":5}]}]},\\"b\\":{\\"$p\\":4}}}"`;

exports[`serialize/deserialize quirky 2`] = `
Object {
  "a": Map {
    Set {
      Object {
        "hello": "world",
        "map": [Circular],
      },
      [Circular],
    } => Object {
      "hello": "world",
      "map": [Circular],
    },
  },
  "b": Set {
    Object {
      "hello": "world",
      "map": [Circular],
    },
    [Circular],
  },
}
`;

exports[`serialize/deserialize 2 quirky 1`] = `"{\\"$r\\":1,\\"$t\\":\\"Object\\",\\"$v\\":{\\"x\\":{\\"$r\\":2,\\"$t\\":\\"Set\\",\\"$v\\":[{\\"$r\\":3,\\"$t\\":\\"Map\\",\\"$v\\":[{\\"$r\\":4,\\"$t\\":\\"Array\\",\\"$v\\":[{\\"$r\\":5,\\"$t\\":\\"Object\\",\\"$v\\":{\\"hello\\":\\"world\\"}},{\\"$r\\":6,\\"$t\\":\\"Map\\",\\"$v\\":[{\\"$r\\":7,\\"$t\\":\\"Array\\",\\"$v\\":[{\\"$p\\":3},{\\"$p\\":5}]},{\\"$r\\":8,\\"$t\\":\\"Array\\",\\"$v\\":[{\\"$p\\":2},{\\"$p\\":5}]}]}]},{\\"$r\\":9,\\"$t\\":\\"Array\\",\\"$v\\":[{\\"$p\\":6},{\\"$p\\":5}]}]},{\\"$p\\":6},{\\"$p\\":5},{\\"x\\":{\\"$p\\":2},\\"a\\":{\\"$p\\":6},\\"b\\":{\\"$p\\":3}},{\\"$p\\":2}]},\\"a\\":{\\"$p\\":6},\\"b\\":{\\"$p\\":3}}}"`;

exports[`serialize/deserialize 2 quirky 2`] = `
Object {
  "a": Map {
    Map {
      Object {
        "hello": "world",
      } => [Circular],
      [Circular] => Object {
        "hello": "world",
      },
    } => Object {
      "hello": "world",
    },
    Set {
      [Circular],
      [Circular],
      [Circular],
      Object {
        "a": [Circular],
        "b": [Circular],
        "x": [Circular],
      },
      [Circular],
    } => Object {
      "hello": "world",
    },
  },
  "b": Map {
    Object {
      "hello": "world",
    } => [Circular],
    [Circular] => Object {
      "hello": "world",
    },
  },
  "x": Set {
    [Circular],
    [Circular],
    Object {
      "hello": "world",
    },
    Object {
      "a": [Circular],
      "b": [Circular],
      "x": [Circular],
    },
    [Circular],
  },
}
`;
